### ISSUE TITLE
Extend SPARQL Protocol to allow limit and offset as query string parameters

### ISSUE THREAD

**Issue by jaw111** [w3c/sparql-12#91]
## Why?

When issuing SPARQL requests programatically, it is a common use case to paginate through the result set by varying the `LIMIT` and `OFFSET`. In such cases it would be handy to be able to keep the query static (as a file) and pass the `limit` and `offset` as:

* query via GET: query string parameters in the request URL
* query via URL-encoded POST: URL-encoded, ampersand-separated query parameters
* query via POST directly: URL-encoded, ampersand-separated query parameters

## Previous work

Dydra supports offset and limit as query string parameters.

## Proposed solution

The HTTP GET request:

```
GET /sparql/?query=PREFIX%20dc%3A%20%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Felements%2F1.1%2F%3E%20%0ASELECT%20%3Fbook%20%3Fwho%20%0AWHERE%20%7B%20%3Fbook%20dc%3Acreator%20%3Fwho%20%7D%0A&limit=100&offset=100 HTTP/1.1
Host: www.example
User-agent: my-sparql-client/0.1
```

Would be equivalent to running the query:

```sparql
PREFIX dc: <http://purl.org/dc/elements/1.1/> 
SELECT ?book ?who 
WHERE { ?book dc:creator ?who }
LIMIT 100
OFFSET 100
```

## Considerations for backward compatibility

None foreseen.

**Comment by kasei** [comment]
@jaw111 Do you think that this should act similarly to dataset specification in the protocol such that protocol-supplied `limit` and `offset` parameters would override any values specified in the query string?

**Comment by namedgraph** [comment]
What about sub-`SELECT`s though?

This use case is addressed straightforwardly with a query builder.

**Comment by kasei** [comment]
@namedgraph I think this is a reasonable feature to consider, and it needn't be concerned with sub-selects. If, as described, you've got a query file on disk, and you want to page through results, this would be a *much* simpler approach than involving any sort of query builder. Similar reasoning could be applied to the dataset construction feature that's already in Protocol 1.1.

**Comment by jaw111** [comment]
@kasei yes, expectation is the protocol-supplied parameters would 'beat' those in the SPARQL query.

**Comment by afs** [comment]
An alternative is that `?limit=` and `?offset=` are valid only if the query string does not contain them.
It meets the paging use case and "static" query example.

**Comment by JervenBolleman** [comment]
@afs this is actually implemented for the HTML reposonses on sparql.uniprot.org. 
When generating a HTML response I found it is always sensible to enforce pagination. (No one wants a 7TB HTML file!) If there is an offset in the query we use it. If there is a LIMIT we take that if it's reasonable for HTML views.

**Comment by cygri** [comment]
Interactive SPARQL clients may want to enforce pagination even for queries where the user has provided an explicit limit. For example, if the user says `LIMIT 1000` but the UI is built to display only 20 rows at a time, then the client still needs to apply a limit of 20 and an appropriate offset as the user pages through the results. Currently, this requires the client to parse/munge the query.

If `&limit=` and `&offset=` were supported in the protocol and would override the value specified in the query, then clients could use the protocol parameters to implement paging, and would not need to modify the user's query. This would simplify the implementation of robust clients.

**Comment by kal** [comment]
@cygri A client would still have to parse the query to know that the user only wants 1000 rows and not page past the end of the block the user asked for. 

That said I would still favour the approach that has the limit and offset parameters in the query string override those of the query for the sake of clarity or simplicity. Otherwise we remain with the question of how the server communicates to the client that a ?limit parameter is not acceptable due to it being present in the query (as opposed to the parameter having an incorrectly formatted value for example).

The can of worms that this starts to open though is the question of a query context supplied to a SPARQL processor by the processor host. This could specify all kinds of different overrides for the query processor to take precendence over what is stated in the query (hence the canned worms). In this case the a web server based processor host might derive the context from a combination of local configuration and request parameters; however you could also imagine a command-line host accepting command line parameters as a way to establish a query context.

**Comment by cygri** [comment]
@kal 

> A client would still have to parse the query to know that the user only wants 1000 rows and not page past the end of the block the user asked for.

If the user wrote `LIMIT 1000`, then the server will not return more than 1000 rows, so there is no possibility of paging past the end. And parsing the query wouldn't tell the client the actual number of results anyway—the query might just have 21 results despite saying `LIMIT 1000`.

**Comment by kal** [comment]
In which case @cygri you are talking about a filter on the result set returned by the query engine, not a parameter that affects the operation of the query engine, which is what I think we are discussing here.

**Comment by cygri** [comment]
@kal Hm. I didn't quite think this through. In my response to you, I assumed the server wouldn't simply ignore the in-query `LIMIT` but honour both the in-query *and* the in-protocol value, and thus not send results beyond the lower of the two limits. But that contradicts what I said earlier (protocol-provided values should override in-query values), and I'm not confident that this is really a good idea, so consider my comment withdrawn.

**Comment by lisp** [comment]
> The can of worms that this starts to open though is the question of a query context supplied to a SPARQL processor by the processor host. This could specify all kinds of different overrides for the query processor to take precendence over what is stated in the query (hence the canned worms). In this case the a web server based processor host might derive the context from a combination of local configuration and request parameters; however you could also imagine a command-line host accepting command line parameters as a way to establish a query context.

this actually indicates how not to make a can of worms out of it:
- enumerate the parameters
- specify their precedence

>  the server wouldn't simply ignore the in-query LIMIT but honour both the in-query and the in-protocol value, and thus not send results beyond the lower of the two limits

this, on the other hand, points the way to a can of worms where, even if combinations are specified the load on the users to get things right could as likely cause them to avoid using the capability.

**Comment by TallTed** [comment]
This seems strongly related to #51.

It seems that `&limit=` would be akin to a LIMIT configured on the server, which over-rides any LIMIT (or lack thereof) in a user's query.  (`&offset=` would be a new thing, but closely related.)

Open questions include how to communicate the presence of such settings/arguments, as well as how to communicate how users may handle their effects.

**Comment by jaw111** [comment]
If it were a `MUST` then no communication necessary beyond which version of SPARQL the endpoint supports.

Otherwise could use SSD and/or Hydra to describe the service.


-------- Original message --------

**Comment by jeenbroekstra** [comment]
As a further data point: limit and offset are also supported in the [RDF4J REST API](https://rdf4j.org/documentation/reference/rest-api/#repository-queries), which the dydra API is based on. In effect other major triplestore vendors (e.g. Ontotext's GraphDB) also support this.

**Comment by TallTed** [comment]
I should have noted this in my previous comment...

`LIMIT` and `OFFSET` are only reliably usable for paging through results when there is also an `ORDER BY` -- as there should be no expectation, because there is no requirement, that the result rows returned for multiple executions of the same query be ordered identically.  (This is the case in SQL as well as SPARQL.)

In other words, you could execute the same query, with the same `LIMIT` and `OFFSET` values, 100 times and receive 100 different orderings of the result rows -- and assuming that the complete result set exceeds the `LIMIT` value, you might even receive a different set of result rows each time!

These values should not be accepted as query string parameters unless `ORDER BY` is *also* added to the list of parameters, the viability of which addition I strongly question. I think that the dydra API and the RDF4J API are likely to cause (indeed, likely to be causing!) problems for their users by supporting `LIMIT` and `OFFSET` parameters unless they also support -- rather, mandate! -- an accompanying `ORDER BY` parameter, which does not appear to be the case. 

Bottom line -- I do not think SPARQL should support `LIMIT` nor `OFFSET` as URI query string parameters.

**Comment by kasei** [comment]
> I think that the dydra API and the RDF4J API are likely to cause (indeed, likely to be causing!) problems for their users by supporting `LIMIT` and `OFFSET` parameters unless they also support -- rather, mandate! -- an accompanying `ORDER BY` parameter, which does not appear to be the case.

I agree that paging in general is problematic without an `ORDER BY`. However, the use of `LIMIT` is not limited to paging. It is enormously helpful during query building/debugging as a way to sanity check expectations against a small set of results. In these cases, requiring ordering at best would impact performance unnecessarily, and at worst might cause a timeout and prevent getting any results back.

**Comment by akuckartz** [comment]
> LIMIT and OFFSET are only reliably usable for paging through results when there is also an ORDER BY ... (This is the case in SQL as well as SPARQL.)

The reference for SPARQL is
https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#modOffset

**Comment by jeenbroekstra** [comment]
> These values should not be accepted as query string parameters unless ORDER BY is also added to the list of parameters, the viability of which addition I strongly question. I think that the dydra API and the RDF4J API are likely to cause (indeed, likely to be causing!) problems for their users by supporting LIMIT and OFFSET parameters unless they also support -- rather, mandate! -- an accompanying ORDER BY parameter, which does not appear to be the case.

This is in fact a known issue in RDF4J (see https://github.com/eclipse/rdf4j/issues/487), though not one that has caused many problems in practice it seems, as we haven't really received any bug reports or requests for improvements along these lines over the years. Most likely this is caused by the fact that database implementations tend to have an implicit stable ordering. While you are quite correct that one shouldn't rely on that, in practice it often "just works".

But that's no basis for a spec design of course. A possible solution that doesn't go as far as mandating an additional parameter (which as you say has its own problems) would be to specify that _if_ a limit or order parameter are specified, the endpoint must take care to use some canonical, repeatable ordering internally. We don't have to specify what that ordering must be, just that consecutive executions will be guaranteed to return the same result in the same order (assuming the data hasn't changed).

**Comment by jaw111** [comment]
Isn't it the case the ORDER BY is not by itself sufficient. The variables that you order by also need to have no two solutions that have the same bindings for those variables in order to guarantee a consistent order.

**Comment by metawilm** [comment]
> But that's no basis for a spec design of course. A possible solution that doesn't go as far as mandating an additional parameter (which as you say has its own problems) would be to specify that _if_ a limit/offset parameter are specified, the endpoint must take care to use some canonical, repeatable ordering internally. We don't have to specify what that ordering must be, just that consecutive executions will be guaranteed to return the same result in the same order (assuming the data hasn't changed).

A specification should focus on semantics, and stay away from introducing implementation constraints that could benefit one product over another. For example in the absence of an `order by` clause it must be acceptable for a query engine to parallelize work, with the eventual ordering of results being non-deterministic.

**Comment by jeenbroekstra** [comment]
> A specification should focus on semantics, and stay away from introducing implementation constraints that could benefit one product over another. For example in the absence of an `order by` clause it must be acceptable for a query engine to parallelize work, with the eventual ordering of results being non-deterministic.

I don't see why saying that when a LIMIT/OFFSET is present an engine imposes an ordering to make the result deterministic is an undue implementation burden, any more than saying that an engine should impose an ordering when an ORDER clause is present. I'm not suggesting that we mandate that the result order is deterministic for _every_ query.

**Comment by jeenbroekstra** [comment]
> Isn't it the case the ORDER BY is not by itself sufficient. The variables that you order by also need to have no two solutions that have the same bindings for those variables in order to guarantee a consistent order.

This is true, though this is not new to the introduction of protocol params: this problem also holds for ORDER BY clauses in queries (in fact it's explicitly mentioned in the spec that behavior is undefined for this case, IIRC). 

I'd be interested to hear of an alternative approach to provide pagination that doesn't have these problems.

**Comment by lisp** [comment]
does it place an "undue burden" on the query author to require that they include the order clause in the query if they intend it to produce stable results when invoked with these request arguments?

**Comment by afs** [comment]
Aside from pagination, #84, there is a case for a basic `?limit=10` which is get a sample of data (or to catch accidentally asking for too much - e.g user/application supplied artitary query).

**Comment by TallTed** [comment]
`LIMIT` without `OFFSET` is indeed useful, and does not require `ORDER BY` to be so.  My bad for not saying so above.  I would be OK with adding `&limit` without `&offset`, with a strong caveat that the spec must then be quite explicit about whether and when `&limit` parameter overrides `LIMIT` clause or vice versa.

> A possible solution that doesn't go as far as mandating an additional parameter ... would be to specify that if a `limit`/`offset` parameter are specified, the endpoint must take care to use some canonical, repeatable ordering internally.

I do not know enough of the internal operations to say with certainty whether this burden would be sufficiently small to be acceptable, but I tend to think it would not.  I think it comes sufficiently close to forcing an implicit `ORDER BY` on every query that this is the burden which must be considered, and while some implementations may be able to do that trivially (because, for instance, they have an internal `ROWID` or similar which can be used for the `ORDER BY`), other implementations may have no option other than "concatenation of all columns in result set" for their `ORDER BY`, which *may* be lightweight but, again, I tend to think not.

**Comment by lisp** [comment]
> LIMIT without OFFSET is indeed useful, and does not require ORDER BY to be so.

likewise offset without limit can provide perhaps useful information independent of order.

**Comment by JervenBolleman** [comment]
I think `limit` is quite simple and worth having in the protocol but for client use cases we should maybe have a different interpretation set in the protocol.And therefore a slightly different name to avoid confusion. Specifically LIMIT for DESCRIBE queries does not act the way I would like for interactive use. With the protocol usage we want to make sure a client does not get so much data in one go that it crashes/aborts/timeouts.

On the other hand OFFSET while one way to paging (currently the only one) might not be what clients really need.
In place of OFFSET/`?offset` maybe we could have an option called `?after`. In a http response we could have a header responding with value that indicates the last row in the binding set.
e.g. given the following query with ?limit=3

```sparqlSELECT ?v 
WHERE { VALUES ?v {1 2 3 4 5}
} ORDER BY ?v
```

Would respond with SPARQL-LAST:3

| v |
| ---: |
| 1 |
| 2 |
| 3 |

Then the same query sent again, now with ?limit=3&after=3 would give

| v |
| ---: |
| 4 |
| 5 |

This after value could be implementation specific. e.g. a query plan hash plus an offset in the plan results, or any other value that would help an implementation give a "reasonable next page".

For clients it is easy to know when it can depend on such a value as the response header would not be present in sparql1.1. servers.

**Comment by jeenbroekstra** [comment]
> `LIMIT` without `OFFSET` is indeed useful, and does not require `ORDER BY` to be so. My bad for not saying so above. I would be OK with adding `&limit` without `&offset`, with a strong caveat that the spec must then be quite explicit about whether and when `&limit` parameter overrides `LIMIT` clause or vice versa.

Good point. Fwiw the current RDF4J spec on this explicitly states that LIMIT/OFFSET in the query overrides the protocol settings. 

> > A possible solution that doesn't go as far as mandating an additional parameter ... would be to specify that if a `limit`/`offset` parameter are specified, the endpoint must take care to use some canonical, repeatable ordering internally.
> 
> I do not know enough of the internal operations to say with certainty whether this burden would be sufficiently small to be acceptable, but I tend to think it would not. I think it comes sufficiently close to forcing an implicit `ORDER BY` on every query that this is the burden which must be considered, and while some implementations may be able to do that trivially (because, for instance, they have an internal `ROWID` or similar which can be used for the `ORDER BY`), other implementations may have no option other than "concatenation of all columns in result set" for their `ORDER BY`, which _may_ be lightweight but, again, I tend to think not.

Fair enough, I think I understand your and @metawilm 's problem better now, and can accept the argument that we should err on the side of 'implementation liberty'.

**Comment by VladimirAlexiev** [comment]
Related: https://github.com/w3c/sparql-12/issues/49. 
I browsed this issue, and didn't see any discussion of cursors or efficiency.
Which is good: those are discussed in #49
